File: test/storage/storage.c
/*
 *  Copyright (C) 2014, 2015, 2016 Michael Hofmann
 *  
 *  This file is part of the Simulation Component and Data Coupling (SCDC) library.
 *  
 *  The SCDC library is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser Public License as published
 *  by the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *  
 *  The SCDC library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser Public License for more details.
 *  
 *  You should have received a copy of the GNU Lesser Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


#include <stdio.h>
#include <stdlib.h>

#include "scdc.h"


#define DBSTORE      0
#define DAVSTORE     0
#define NFSSTORE     0
#define TARGET       "store_test"
#define INPUT_FILE   TARGET ".in"
#define OUTPUT_FILE  TARGET ".out"


int main(int argc, char *argv[])
{
  const char *uri = 
#if DBSTORE
    "scdc:/storeDB";
#elif DAVSTORE
    "scdc:/storeDAV";
#elif NFSSTORE
    "scdc:/storeNFS";
#else
    "scdc:/storeFS";
#endif
  const char *cmd;
  char path[256], *e;
  scdc_dataprov_t dpFS, dpDB, dpDAV, dpNFS;
  scdc_dataset_t ds;
  scdc_dataset_input_t _ip, *ip = &_ip;
  scdc_dataset_output_t _op, *op = &_op;


  printf("SCDC storage demo for C\n");

  scdc_init(SCDC_INIT_DEFAULT);

  dpFS = scdc_dataprov_open("storeFS", "fs", (e = getenv("MERGE_SCDC_REPO_PATH"), sprintf(path, "%s%s", (e?e:""), "store/"), path));
  dpDB = scdc_dataprov_open("storeDB", "mysql", getenv("MERGE_SCDC_MYSQL_CREDENTIALS"));
  dpDAV = scdc_dataprov_open("storeDAV", "webdav");
  dpNFS = scdc_dataprov_open("storeNFS", "nfs");


  /* open dataset */
  ds = scdc_dataset_open(uri);
  if (ds == SCDC_DATASET_NULL)
  {
    printf("ERROR: open dataset failed!\n");
    goto quit;

  } else printf("open dataset '%s': OK\n", uri);

#if DBSTORE
  /* become admin on db service */
  cmd = "cd admin";
  if (scdc_dataset_cmd(ds, cmd, NULL, NULL) != SCDC_SUCCESS)
  {
    printf("ERROR: command '%s' failed!\n", cmd);
    goto quit;

  } else printf("becoming admin with command '%s': OK\n", cmd);

  /* remove storage table (do not care whether it fails or not) */
  cmd = "rm storage_demo";
  scdc_dataset_cmd(ds, cmd, NULL, NULL);

  /* add storage table */
  cmd = "put storage_demo";
  if (scdc_dataset_cmd(ds, cmd, NULL, NULL) != SCDC_SUCCESS)
  {
    printf("ERROR: command '%s' failed!\n", cmd);
    goto quit;

  } else printf("adding data base storage table with command '%s': OK\n", cmd);

  /* add storage table */
  cmd = "cd storage_demo";
  if (scdc_dataset_cmd(ds, cmd, NULL, NULL) != SCDC_SUCCESS)
  {
    printf("ERROR: command '%s' failed!\n", cmd);
    goto quit;

  } else printf("selecting data base storage table with command '%s': OK\n", cmd);
#endif

  /* create input object to read from file */
  ip = scdc_dataset_input_create(ip, "file", INPUT_FILE);
  if (ip == NULL)
  {
    printf("ERROR: open input file failed!\n");
    goto quit;
  }
    
  /* store data from input object to dataset */
  cmd = "put " TARGET;
  if (scdc_dataset_cmd(ds, cmd, ip, NULL) != SCDC_SUCCESS)
  {
    printf("ERROR: command '%s' failed!\n", cmd);
    goto quit;

  } else printf("storing data with command '%s': OK\n", cmd);

  /* destroy input object */
  scdc_dataset_input_destroy(ip);

  /* create output object to write to file */
  op = scdc_dataset_output_create(op, "file", OUTPUT_FILE);
  if (op == NULL)
  {
    printf("ERROR: open output file failed!\n");
    goto quit;
  }

  /* retrieve data from dataset to output object */
  cmd = "get " TARGET;
  if (scdc_dataset_cmd(ds, cmd, NULL, op) != SCDC_SUCCESS)
  {
    printf("ERROR: command '%s' failed!\n", cmd);
    goto quit;

  } else printf("retrieving data with command '%s': OK\n", cmd);

  /* destroy output object */
  scdc_dataset_output_destroy(op);

  /* remove the data stored */
  cmd = "rm " TARGET;
  if (scdc_dataset_cmd(ds, cmd, NULL, NULL) != SCDC_SUCCESS)
  {
    printf("ERROR: command '%s' failed!\n", cmd);
    goto quit;

  } else printf("removing data with command '%s': OK\n", cmd);

#if DBSTORE
  /* become admin on db service */
  cmd = "cd admin";
  if (scdc_dataset_cmd(ds, cmd, NULL, NULL) != SCDC_SUCCESS)
  {
    printf("ERROR: command '%s' failed!\n", cmd);
    goto quit;

  } else printf("becoming admin with command '%s': OK\n", cmd);

  /* remove storage table (do not care whether it fails or not) */
  cmd = "rm storage_demo";
  scdc_dataset_cmd(ds, cmd, NULL, NULL);
#endif

  /* close dataset */
  scdc_dataset_close(ds);

quit:
  scdc_dataprov_close(dpFS);
  scdc_dataprov_close(dpDB);
  scdc_dataprov_close(dpDAV);
  scdc_dataprov_close(dpNFS);

  scdc_release();

  return 0;
}

File: test/storage.py/storage.py
#!/usr/bin/python

#  
#  Copyright (C) 2014, 2015, 2016, 2017 Michael Hofmann
#  
#  This file is part of the Simulation Component and Data Coupling (SCDC) library.
#  
#  The SCDC library is free software: you can redistribute it and/or
#  modify it under the terms of the GNU Lesser Public License as published
#  by the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  
#  The SCDC library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Lesser Public License for more details.
#  
#  You should have received a copy of the GNU Lesser Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
#  


import os
import sys
import scdc


#print("path: ", sys.path)

STORE_FS_PATH = os.getenv("MERGE_SCDC_REPO_PATH", "") + "store/"
STORE_MYSQL_CREDENTIALS = os.getenv("MERGE_SCDC_MYSQL_CREDENTIALS", "")


dbstore = False
target = "store_test"
input_file = target + ".in"
output_file = target + ".out"


print("SCDC storage demo for Python");

scdc.init()

dpFS = scdc.dataprov_open("storeFS", "fs", STORE_FS_PATH)
dpDB = scdc.dataprov_open("storeDB", "mysql", STORE_MYSQL_CREDENTIALS)

uri = "scdc:/storeDB" if dbstore else "scdc:/storeFS"


# the loop is only used to be able to break if an error occurs
while True:

  # open dataset
  ds = scdc.dataset_open(uri)
  if not ds:
    print("ERROR: open dataset failed!")
    break
  else:
    print("open dataset '" + uri + "': OK")

  if dbstore:
    # become admin on db service
    cmd = "cd admin"
    if not scdc.dataset_cmd(ds, cmd, None, None):
      print("ERROR: command '" + cmd +"' failed!")
      break
    else:
      print("becoming admin with command '" + cmd + "': OK")

    # remove storage table (do not care whether fails or not)
    cmd = "rm storage_demo"
    scdc.dataset_cmd(ds, cmd, None, None)

    # add storage table
    cmd = "put storage_demo"
    if not scdc.dataset_cmd(ds, cmd, None, None):
      print("ERROR: command '" + cmd +"' failed!")
      break
    else:
      print("adding data base storage table with command '" + cmd + "': OK")

    # add storage table
    cmd = "cd storage_demo"
    if not scdc.dataset_cmd(ds, cmd, None, None):
      print("ERROR: command '" + cmd +"' failed!")
      break
    else:
      print("selecting data base storage table with command '" + cmd + "': OK")

  # create input object to read from file
  ip = scdc.dataset_input_create("file", input_file)
  if not ip:
    print("ERROR: open input file failed!")
    break
    
  # store data from input object to dataset
  cmd = "put " + target
  if not scdc.dataset_cmd(ds, cmd, ip, None):
    print("ERROR: command '" + cmd + "' failed!")
    break
  else:
    print("storing data with command '" + cmd + "': OK")

  # destroy input object
  scdc.dataset_input_destroy(ip)

  # create output object to write to file
  op = scdc.dataset_output_create("file", output_file)
  if not op:
    print("ERROR: open output file failed!")
    break

  # retrieve data from dataset to output object
  cmd = "get " + target
  if not scdc.dataset_cmd(ds, cmd, None, op):
    print("ERROR: command '" + cmd + "' failed!")
    break
  else:
    print("retrieving data with command '" + cmd + "': OK")

  # destroy output object
  scdc.dataset_output_destroy(op)

  # remove the data stored
  cmd = "rm " + target
  if not scdc.dataset_cmd(ds, cmd, None, None):
    print("ERROR: command '" + cmd + "' failed!")
    break
  else:
    print("removing data with command '" + cmd + "': OK")

  if dbstore:
    # become admin on db service
    cmd = "cd admin"
    if not scdc.dataset_cmd(ds, cmd, None, None):
      print("ERROR: command '" + cmd +"' failed!")
      break
    else:
      print("becoming admin with command '" + cmd + "': OK")

    # remove storage table (do not care whether fails or not)
    cmd = "rm storage_demo"
    scdc.dataset_cmd(ds, cmd, None, None)

  # close dataset
  scdc.dataset_close(ds);

  # always quit the loop and the end
  break


scdc.dataprov_close(dpFS)
scdc.dataprov_close(dpDB)
